### PR TITLE
Surface normals and wireframe

### PR DESCRIPTION
**Pull request by brisvag**
# Description

Add the ability to display edge wireframe, face normals and vertex normals for the `Surface` layer.

Example:

https://user-images.githubusercontent.com/23482191/143427532-e98cd62d-1593-491b-a029-01784d9b724a.mp4

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References
- #3669
- #3334
- #2970

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/_vispy/layers/surface.py
Comment (brisvag) [inline napari/_vispy/layers/surface.py:39]: Most mesh software follows a convention on the order of vertices to determine whether the normal faces in or out. This `[:, ::-1]` Was reversing the order of the points, thus inverting all the normals. Does anyone know why this was introduced? To me, it looks like it might be a copy-paste leftover from the `Points` layer, where reversing is used to diplay the latest added point on top.

Comment (kevinyamauchi) [reply napari/_vispy/layers/surface.py:39]: Is this reversing the order of the vertices or swapping the dims order of the points? I think it might be the later. If so, I think this is because we store thing zyx in napari and vispy expects xyz.

Comment (kevinyamauchi) [reply napari/_vispy/layers/surface.py:39]: I am pretty sure the order of the vertices in a give faces is set by `self.layer._view_faces` below.

**Comment by kevinyamauchi** [comment]
I am excited for this! Thanks, @brisvag !

File: examples/surface_normals_wireframe.py
Comment (alisterburt) [inline examples/surface_normals_wireframe.py]:
```suggestion
vert *= 100  # c.f. https://github.com/napari/napari/issues/3477
```

**Review by alisterburt — COMMENTED** [review]
gorgeous! I wonder if this might be a case where we reduce the flat-ness of the layer API and bundle things together in an `EventedModel`

I'm imagining `Surface.wireframe`, `Surface.face_normals` and `Surface.vertex_normals` being instances of `Wireframe` and `Vectors` models - thoughts @brisvag @napari/core-devs ? I know flat API has been something napari has aimed to keep for a while so wouldn't want this to happen without discussion

File: napari/_vispy/layers/surface.py
Comment (brisvag) [reply napari/_vispy/layers/surface.py:39]: Uh-oh, got confused! Thanks :)

**Comment by brisvag** [comment]
Yup, I thought about using `EventedModel` too, this layer's api got very wide very quickly.

Ny the way, something that should be said about the current implementation: it should probably be changed to only create the `normal` visuals if they are displayed. Right now, there's a lot of data being created and shuffled around for no reason even if the user never uses this functionality.

**Comment by alisterburt** [comment]
Nice, do you see that as blocking for this PR or would you rather follow up with that in a second PR soon?

**Comment by brisvag** [comment]
Not really blocking, but might be easy to fix here already :)

File: napari/_vispy/layers/surface.py
Comment (brisvag) [reply napari/_vispy/layers/surface.py:39]: Should be fixed now: I reverse the faces definition as well, which should restore handedness after the `xyz` to `zyx` change.

**Comment by brisvag** [comment]
> Not really blocking, but might be easy to fix here already :)

Done :) I did not change to to "purge" the normals if one disables them, with the reasoning that if you enable them once, you probably want them again later, and memory usage is (probably) less valuable than speed of enable/disable.

File: napari/_vispy/layers/surface.py
Comment (alisterburt) [reply napari/_vispy/layers/surface.py:39]: should be or is? :) if this is fixed we should be good to merge now!

Comment (brisvag) [reply napari/_vispy/layers/surface.py:39]: "should" as in "I hope I didn't break something else" :P

On a side note, I think we should wait for others to chip in before we merge; this PR even more than #3430 has a lot of new API, so there's probably some talk to have there. Also: should this PR include GUI access to this stuff?

File: napari/layers/surface/surface.py
Comment (sofroniewn) [inline napari/layers/surface/surface.py]: As mentioned above, I think it would be nice to put these all in their own EventedModels - purely for encapsulation reasons. Could do 

```python
layer.wireframe.enabled
layer.wireframe.width
layer.wireframe.color

# and

layer.normals.face.enabled
layer.normals.face.color
layer.normals.face.width
layer.normals.face.length

# and 

layer.normals.vertex.enabled
layer.normals.vertex.color
layer.normals.vertex.width
layer.normals.vertex.length
```

That would probably be my top choice API. What does @alisterburt @kevinyamauchi @jni @alisterburt think?

Comment (brisvag) [reply napari/layers/surface/surface.py]: Yeah I agree!

**Comment by brisvag** [comment]
Switched over to `EventedModel`. Things I'm in doubt about:
- type hints
- where the new models live

Other than that, it works!

File: napari/layers/surface/surface.py
Comment (sofroniewn) [inline napari/layers/surface/surface.py]: Getting fails

```python-traceback
  E       AttributeError: 'SurfaceWireframe' object has no attribute 'to_dict'
```

Should this just be `self.wireframe.dict()`? I can't remember off top of my head

**Review by sofroniewn — COMMENTED** [review]
Location of files LGTM. I flagged one failing test.

Maybe @tlambert03 can advise on type hints. Fine by me to merge as is though.

Once tests are fixed we can get in!!

**Review by jni — APPROVED** [review]

File: napari/layers/surface/surface.py
Comment (brisvag) [reply napari/layers/surface/surface.py]: Fixed, which however made something else clear: this cannot work since `wireframe` is no longer an argument of `points`. should it still be an argument, and accept `dict | SurfaceWireframe` as types? (EDIT: that's what I went with for now)

Comment (alisterburt) [reply napari/layers/surface/surface.py]: Yep that sounds good! It's what we did for Slicing plane and co 🙂

**Review by alisterburt — APPROVED** [review]

File: napari/layers/surface/wireframe.py
Comment (tlambert03) [inline napari/layers/surface/wireframe.py:24]: really wish we could somehow standardize this :/
we accept colors _so_ many places... it would be great if we didn't still have to define them as unions of various valid inputs.  (among other things, this means that, as far as the typing implies, the _consumer_ of `wireframe.color` needs to be able to handle any of these options).

if you can't use `pydantic.color.Color`, maybe you can start to think about a napari color field?  @andy-sweet has also looked into this and knows a lot about the sneaky pitfalls.  but just pinning it again.

Comment (andy-sweet) [reply napari/layers/surface/wireframe.py:24]: I usually think about the declared type as the one that will be stored. For a single color value in napari, I think that's usually `Array[float, (4,)]` (and is almost the type returned by `transform_color` - that actually has shape (-1, 4)).

For supported parameter/inputs, we have `ColorType`, but that includes vispy's `Color`, which isn't currently supported by `transform_color` (though could be easily).

I started writing a developer guide around colors in napari, which might be helpful to understand some of the details here: https://hackmd.io/K5rCILaHQmWsNhY0ml_xaQ

In the style encoding work, I defined a custom types for a single `ColorValue` and multiple `ColorArray`: https://github.com/napari/napari/pull/3493/files#diff-c6c515811f47fb2a55c3edb7f778cc93b0bbf1435d21c11a11e78a3d6d1e7ba5R23

I'd be nervous about bringing in pydantic's `Color` at this point (even though it's great!), because it would add some inconsistency (e.g. it supports CSS3 color names, but it doesn't support matplotlib's single character names).

Comment (brisvag) [reply napari/layers/surface/wireframe.py:24]: I agree that this is not good; is this something you consider blocking for this PR and we should work on first?

Comment (alisterburt) [reply napari/layers/surface/wireframe.py:24]: I don't think this feature should be blocked until colours are standardised, it would presumably be a fairly easy thing to update once progress was made on colours?

File: napari/layers/surface/surface.py
Comment (brisvag) [reply napari/layers/surface/surface.py]: Yeah that's where I looked for inspiration :)

**Review by alisterburt — APPROVED** [review]

File: napari/layers/surface/wireframe.py
Comment (tlambert03) [reply napari/layers/surface/wireframe.py:24]: Yeah, definitely not blocking

**Review by sofroniewn — APPROVED** [review]
Looks great! Will merge now!!

**Pull request merged**
